### PR TITLE
[IIIF-1121] Handle access token requests

### PR DESCRIPTION
**Pull request by markmatney**
Summary:

- Handle access token requests
- Add new env var ACCESS_TOKEN_EXPIRES_IN for configuring the expiry of access tokens (which defaults to 3600 seconds)
- Add AccessCookieServiceError enum to match the pattern used by DatabaseService (DatabaseServiceError)
- Move path template strings used in ITs to AbstractHandlerIT
- Update OpenAPI spec
- See TokenJsonKeys for details of access token structure

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,6 +36,7 @@ These are a work in progress. None of these need to be supplied when running in
 | --- | --- | --- |
 | HTTP_PORT | 8888 | No |
 | HTTP_HOST | 0.0.0.0 | No |
+| ACCESS_TOKEN_EXPIRES_IN | 3600 | No |
```
Comment (DRickard) [inline README.md:39]: This is set as 3600 here, but 1800 in pom (lines 410, 520)

File: src/main/java/edu/ucla/library/iiif/auth/handlers/AccessTokenHandler.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    /**
+     * Creates a handler that exchanges access cookies for access tokens.
+     *
+     * @param aVertx The Vert.x instance
+     * @param aConfig A configuration
+     */
+    public AccessTokenHandler(final Vertx aVertx, final JsonObject aConfig) {
+        myConfig = aConfig;
+        myExpiresIn = aConfig.getInteger(Config.ACCESS_TOKEN_EXPIRES_IN, 3600);
```
Comment (DRickard) [inline src/main/java/edu/ucla/library/iiif/auth/handlers/AccessTokenHandler.java:62]: And here the expiration is 3600

File: README.md (hunk 1 above)
Comment (markmatney) [reply README.md:39]: I set it to 1800 in the POM just to show that the value is configurable -- the default value for the application is intended to be 3600.

Comment (DRickard) [reply README.md:39]: Ah...

File: src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoService.java [hunk 3]
```diff
@@ -27,30 +23,16 @@
...
      */
     @SuppressWarnings({ "PMD.PreserveStackTrace" })
-    static AccessCookieCryptoService create(final JsonObject aConfig) {
-        try {
-            return new AccessCookieCryptoServiceImpl(aConfig);
-        } catch (final InvalidKeySpecException | NoSuchAlgorithmException | NoSuchPaddingException details) {
-            throw new ServiceException(CONFIGURATION_ERROR, details.getMessage());
-        }
+    static AccessCookieCryptoService create(final JsonObject aConfig) throws GeneralSecurityException {
+        return new AccessCookieCryptoServiceImpl(aConfig);
```
Comment (ksclarke) [inline src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoService.java:35]: Why not AccessCookieService and AccessCookieServiceImpl, etc. Is the crypto important enough here to include in the name?

File: src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoServiceError.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package edu.ucla.library.iiif.auth.services;
+
+/**
+ * AccessCookieCryptoService errors.
+ */
+public enum AccessCookieCryptoServiceError {
```
Comment (ksclarke) [inline src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoServiceError.java:6]: Same question about "crypto" in the name as above.

File: src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoServiceImpl.java [hunk 5]
```diff
@@ -134,12 +162,20 @@
...
-        final byte[] encryptedCookieData = decodedCookie.getBinary(CookieJsonKeys.SECRET);
-        final byte[] nonce = decodedCookie.getBinary(CookieJsonKeys.NONCE);
+        final JsonObject decodedCookie;
+        final byte[] encryptedCookieData;
+        final byte[] nonce;
         final byte[] serializedCookieData;
         final JsonObject cookieData;
 
+        try {
+            decodedCookie = new JsonObject(new String(Base64.getDecoder().decode(aCookieValue)));
```
Comment (ksclarke) [inline src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoServiceImpl.java:172]: Could this `decodedCookie` variable be defined in the try scope? If it's not used outside of it, it seems cleaner to keep it isolated to that scope.

File: src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoService.java (hunk 3 above)
Comment (markmatney) [reply src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieCryptoService.java:35]: Agreed, will rename

**Review by ksclarke — COMMENTED** [review]
Looks good. Just had a couple of questions handled as individual comments.

**Comment by ksclarke** [comment]
Regarding "See commit messages." for the PR message... it still would be nice to summarize for us :-)

File: src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieServiceImpl.java [hunk 6]
```diff
@@ -134,24 +162,36 @@
...
-        final byte[] encryptedCookieData = decodedCookie.getBinary(CookieJsonKeys.SECRET);
-        final byte[] nonce = decodedCookie.getBinary(CookieJsonKeys.NONCE);
-        final byte[] serializedCookieData;
         final JsonObject cookieData;
 
         try {
+            final JsonObject decodedCookie = new JsonObject(new String(Base64.getDecoder().decode(aCookieValue)));
+            final byte[] encryptedCookieData = decodedCookie.getBinary(CookieJsonKeys.SECRET);
+            final byte[] nonce = decodedCookie.getBinary(CookieJsonKeys.NONCE);
+            final byte[] serializedCookieData;
```
Comment (markmatney) [inline src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieServiceImpl.java:171]: I realize this is a bit more than you asked for @ksclarke but I think it's an improvement (?)

Comment (ksclarke) [reply src/main/java/edu/ucla/library/iiif/auth/services/AccessCookieServiceImpl.java:171]: Looks good to me.

**Review by ksclarke — APPROVED** [review]

**Comment by markmatney** [comment]
The test run found an edge case of the JSON of a tampered cookie still being syntactically valid, so I added another commit to make the check more thorough.

**Review by ksclarke — APPROVED** [review]
👍🏼

**Review by DRickard — APPROVED** [review]
